src/tests/adr/vla_adr.rs

```diff
@@ -0,0 +1,503 @@
+            },
+        }
+    "###);
+
+    // ...an array of type DINT with its dimensions unknown at compile time
+    insta::assert_debug_snapshot!(index.find_effective_type_by_name("__arr_vla_1_dint").unwrap(),
+    @r###"
+    DataType {
+        name: "__arr_vla_1_dint",
+        initial_value: None,
+        information: Array {
+            name: "__arr_vla_1_dint",
+            inner_type_name: "DINT",
+            dimensions: [
+                Dimension {
+                    start_offset: Undetermined,
+                    end_offset: Undetermined,
+                },
+            ],
+        },
+        nature: Any,
+        location: SymbolLocation {
+            line_number: 4294967295,
+            source_range: SourceRange {
+                range: 0..0,
+            },
+        },
+    }
+    "###);
+
+    // Finally the dimensions array, which is being populated at runtime; see [`pass`]
+    insta::assert_debug_snapshot!(index.find_effective_type_by_name("n_dims").unwrap(), 
+    @r###"
+        DataType {
+            name: "n_dims",
+            initial_value: None,
+            information: Array {
+                name: "n_dims",
+                inner_type_name: "DINT",
+                dimensions: [
+                    Dimension {
+                        start_offset: ConstExpression(
+                            Index {
+                                index: 0,
+                                generation: 0,
+                            },
+                        ),
+                        end_offset: ConstExpression(
+                            Index {
+                                index: 1,
+                                generation: 0,
+                            },
+                        ),
+                    },
+                ],
+            },
+            nature: Any,
+            location: SymbolLocation {
+                line_number: 0,
+                source_range: SourceRange {
+                    range: 0..0,
+                },
+            },
+        }
+    "###);
+}
+
+/// Because VLAs are internally handled as structs, they'll naturally also translate into LLVM structs.
+#[test]
+fn declare() {
+    let src = "
+        FUNCTION foo : DINT
+            VAR_INPUT {ref}
+                arr : ARRAY[*] OF DINT;
+            END_VAR
+        END_FUNCTION
+    ";
+
+    assert!(codegen(src).contains("%__foo_arr = type { i32*, [2 x i32] }"));
+}
+
+/// VLAs (in RuSTy) are defined to be always by-ref, meaning POUs accepting VLAs expect a pointer to a struct.
+/// To satisfy that constraint, any array passed to such a POU needs to be first wrapped inside a struct
+/// with their fields populated. Therefore, whenever we encounter such situations we first stack-allocate
+/// a struct then pass (1) the array's address and (2) an array consisting of the dimensions'
+/// {start,end}-offset.
+/// XXX: It would be ideal to only stack allocate once instead of allocating with every POU call
```

Comment:
I get your point but stack allocation is the cheapest thing you can do, and your allocation is more or less small  (1 pointer plus a couple of ints), and LLVM passes could spot this. - So i would not bother further!

Reply:
That comment is a bit misleading, my bad. Not only do we call `alloca` with every function call but actually generate all IR related logic within the `Castable` impl of `ArrayValue`; i.e. multiple duplicate instructions with each function call, see: https://github.com/PLC-lang/rusty/blob/a9a04a3bd8bd27ca744d94cecb9d0c236b7eedb0/src/codegen/llvm_typesystem.rs#L220-L278 Michael and I also internally discussed about this and came to the conclusion that we ignore it for now hoping the LLVM optimizer deals with it. Now that I think about it we also wanted to verify whether or not this is actually the case :P

Reply:
I agree with your conclusion